piker/brokers/kucoin.py

```diff
@@ -0,0 +1,670 @@
+                                'response': True,
+                            }
+                        )
+
+            async with (
+                open_autorecon_ws(
+                    f'wss://ws-api-spot.kucoin.com/?token={token}&[connectId={connect_id}]',
+                    fixture=subscribe,
+                ) as ws,
+                stream_messages(ws, sym) as msg_gen,
+            ):
+                typ, quote = await anext(msg_gen)
+
+                while typ != 'trade':
+                    # TODO: use ``anext()`` when it lands in 3.10!
+                    typ, quote = await anext(msg_gen)
+
+
+                task_status.started((init_msgs, quote))
+                feed_is_live.set()
+
+                async for typ, msg in msg_gen:
+                    await send_chan.send({sym: msg})
+
+
+def make_sub(sym, connect_id, level='l1') -> dict[str, str | bool] | None:
+    match level:
+        case 'l1':
+            return {
+                'id': connect_id,
+                'type': 'subscribe',
+                'topic': f'/spotMarket/level2Depth5:{sym}',
+                'privateChannel': False,
+                'response': True,
+            }
+
+        case 'l3':
+            return {
+                'id': connect_id,
+                'type': 'subscribe',
+                'topic': f'/market/ticker:{sym}',
+                'privateChannel': False,
+                'response': True,
+            }
+
+@trio_async_generator
+async def stream_messages(ws: NoBsWs, sym: str) -> AsyncGenerator[NoBsWs, dict]:
+    timeouts = 0
+    last_trade_ts = 0
+
+    while True:
+        with trio.move_on_after(3) as cs:
+            msg = await ws.recv_msg()
+        if cs.cancelled_caught:
+            timeouts += 1
+            if timeouts > 2:
+                log.error('kucoin feed is sh**ing the bed... rebooting...')
+                await ws._connect()
+
+            continue
+
+
+        if msg.get('subject') != None:
+            msg = KucoinMsg(**msg)
+            match msg.subject:
+                case 'trade.ticker':
+                    trade_data = KucoinTrade(**msg.data)
+
+                    # XXX: Filter for duplicate messages as ws feed will send duplicate market state
+                    # https://docs.kucoin.com/#level2-5-best-ask-bid-orders
+                    if trade_data.time == last_trade_ts:
+                        continue
+
+                    last_trade_ts = trade_data.time
+
+                    yield 'trade', {
+                        'symbol': sym,
+                        'last': trade_data.price,
+                        'brokerd_ts': trade_data.time,
+                        'ticks': [
+                            {
+                                'type': 'trade',
+                                'price': float(trade_data.price),
+                                'size': float(trade_data.size),
+                                'broker_ts': trade_data.time,
```

Comment:
why not use the `last_trade_ts` var you just set to avoid yet another attr lookup.